### ISSUE TITLE
Caffe2 build breaks other unit tests due to protocol buffer version conflict

### ISSUE THREAD

**Issue by beniz** [jolibrain/deepdetect#447]
How to reproduce:

- build with caffe2 + xgboost

```
cmake .. -DBUILD_TESTS=ON -DUSE_CAFFE2=ON -DUSE_XGBOOST=ON 
```

- run `./ut_caffe_mlp` in `tests`:

```
[ RUN      ] caffelib.configure_resnet_template_n_nt
[2018-06-20 10:52:01.774] [test] [info] generating ResNet with depth=50 / n=5
[libprotobuf ERROR /home/beniz/tmp/deepdetect/build_caffe2/pytorch/src/pytorch/third_party/protobuf/src/google/protobuf/descriptor.cc:3592] Invalid proto descriptor for file "caffe.proto":
[libprotobuf ERROR /home/beniz/tmp/deepdetect/build_caffe2/pytorch/src/pytorch/third_party/protobuf/src/google/protobuf/descriptor.cc:3595]   caffe.proto: Unrecognized syntax: Softmax
[libprotobuf FATAL /home/beniz/tmp/deepdetect/build_caffe2/pytorch/src/pytorch/third_party/protobuf/src/google/protobuf/generated_message_reflection.cc:2329] CHECK failed: file != NULL: 
unknown file: Failure
C++ exception with description "CHECK failed: file != NULL: " thrown in the test body.
[  FAILED  ] caffelib.configure_resnet_template_n_nt (10 ms)
```

**Comment by chichaj** [comment]
Possible fix on [caffe/pull/24](https://github.com/jolibrain/caffe/pull/24)

**Comment by beniz** [comment]
Thanks for the script in https://github.com/jolibrain/caffe/pull/24 it's been very useful! 

I've tested with several versions of protobuf and I was eventually able to trace the source back to a Caffe2 building option that prevents the building of protobuf shared libraries.

The solution appears to force `CAFFE2_LINK_LOCAL_PROTOBUF` to 0. I'm not sure how to do this through the regular options to cmake, so for testing purposes I've forced the value from within the cmake files. Below is my diff to an up to date version of Pytorch:

```
diff --git a/cmake/ProtoBuf.cmake b/cmake/ProtoBuf.cmake
index c270471..4521000 100644
--- a/cmake/ProtoBuf.cmake
+++ b/cmake/ProtoBuf.cmake
@@ -10,6 +10,7 @@ macro(custom_protobuf_find)
     # so we turn it off here.
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-declarations")
   endif()
+  set(CAFFE2_LINK_LOCAL_PROTOBUF "0")
   if (${CAFFE2_LINK_LOCAL_PROTOBUF})
     # If we are going to link protobuf locally, we will need to turn off
     # shared libs build for protobuf.
```

Rebuilding caffe2 from scratch allows your bash script to succeed in all cases.

From a quick test you can obtain the same result with:

```
cmake .. -DCAFFE2_LINK_LOCAL_PROTOBUF=0
```

Now, compiling DD with the following patch:

```diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9c6cbc3..44c9be4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -132,6 +132,8 @@ if (USE_CAFFE2)
 
     -DCMAKE_PREFIX_PATH=${EIGEN3_INSTALL}
 
+    -DCAFFE2_LINK_LOCAL_PROTOBUF=0
+    
     -DUSE_OPENMP=ON
     -DUSE_MPI=OFF
```
allows to build the shared protobuf libs from within Pytorch/Caffe2.

However, this brings another problem, that `libcaffe` and `libcaffe2` now appear to link to the compiled caffe protobuf twice:

```
[libprotobuf ERROR /home/beniz/code/deepdetect/build/pytorch/src/pytorch/third_party/protobuf/src/google/protobuf/descriptor_database.cc:111] Symbol name "caffe.BlobShape" conflicts with the existing symbol "caffe.BlobShape".
[libprotobuf FATAL /home/beniz/code/deepdetect/build/pytorch/src/pytorch/third_party/protobuf/src/google/protobuf/descriptor.cc:1394] CHECK failed: generated_database_->Add(encoded_file_descriptor, size): 
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  CHECK failed: generated_database_->Add(encoded_file_descriptor, size): 
Aborted (core dumped)
```

This has been reported before: https://github.com/caffe2/caffe2/issues/2457

I believe the way to solve this is to build the caffe.proto only once. One idea could be to replace the `pytorch/caffe/proto/` directory by a symlink to the `caffe/src/proto/` dir and then find a way to only build it once and not include it within the two dynamic libraries.

I am not committing the last patch above since this breaks all the exe of a DD / caffe2 build....

**Comment by beniz** [comment]
A quick check shows that `caffe::Datum` is used by caffe2 as a compatibility with data sources used with Caffe. Since this is useful to us, disabling caffe.proto in caffe2 is not an option. We should then find a way to work with both!

Example: 

`caffe2/video/video_io.h` has ```#include <caffe/proto/caffe.pb.h>```

**Comment by chichaj** [comment]
Here is an fix tentative #452

**Comment by beniz** [comment]
Thanks.
